server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/mailet/filter/ActionApplier.java

```diff
@@ -127,21 +128,33 @@ private void applyReject(Rule.Action action) {
     /**
      * @return a boolean value to show if a local copy should be kept or not
      */
-    private boolean applyForward(Rule.Action action) {
-        return action.getForward().map(Throwing.function(forward -> {
-            LoopPrevention.RecordedRecipients recordedRecipients = LoopPrevention.RecordedRecipients.fromMail(mail);
-            Set<MailAddress> newRecipients = recordedRecipients.nonRecordedRecipients(forward.getAddresses());
-            boolean shouldMailBeForwarded = !newRecipients.isEmpty();
-            if (shouldMailBeForwarded) {
-                sendACopy(mailetContext, mailAddress, recordedRecipients, newRecipients);
-            }
-            if (!forward.isKeepACopy() && shouldMailBeForwarded) {
-                removeFromRecipients();
-                return false;
-            } else {
-                return true;
-            }
-        })).orElse(false);
+    private void applyForward(Rule.Action action) {
+        action.getForward()
+            .filter(any -> shouldProcessingContinue())
+            .ifPresent(Throwing.consumer(forward -> {
+                LoopPrevention.RecordedRecipients recordedRecipients = LoopPrevention.RecordedRecipients.fromMail(mail);
+
+                if (recordedRecipients.getRecipients().contains(mailAddress)) {
+                    // Forward is already processed. Do not do it again.
+                    return;
+                }
+
+                Set<MailAddress> newRecipients = getNewRecipients(forward, recordedRecipients);
+                if (!newRecipients.isEmpty()) {
+                    removeFromRecipients();
+                    sendACopy(mailetContext, mailAddress, recordedRecipients, newRecipients);
+                }
+            }));
+    }
+
+
+    private Set<MailAddress> getNewRecipients(Rule.Action.Forward forward, LoopPrevention.RecordedRecipients recordedRecipients) {
+        ImmutableSet.Builder<MailAddress> newRecipientsBuilder = ImmutableSet.<MailAddress>builder()
+            .addAll(recordedRecipients.nonRecordedRecipients(forward.getAddresses()));
+        if (forward.isKeepACopy()) {
+            newRecipientsBuilder.add(mailAddress);
```

Comment:
Current user forward email to himself??

Reply:
As explained, loclCopy is now handled by keeping the original recipient on the forwarded mail: easier...